crates/transpiler/src/passes/commutative_optimization.rs

```
@@ -0,0 +1,516 @@
+            let qargs = dag.get_qargs(inst.qubits);
+            if !qargs.is_sorted() {
+                let mut sorted_qargs = qargs.to_vec();
+                sorted_qargs.sort();
+                let sorted_qubits = dag.add_qargs(&sorted_qargs);
+                let canonical_instruction = PackedInstruction::from_standard_gate(
+                    standard_gate,
+                    inst.params.clone(),
+                    sorted_qubits,
+                );
+                return Some((canonical_instruction, Param::Float(0.)));
+            }
+        }
+    }
+    None
+}
+
+/// Return `true` if two instructions commute (up to the specified tolerance).
+///
+/// # Arguments:
+///
+/// * `dag`: The output [DAGCircuit] that contains all interned qubits.
+/// * `approximation_degree`: Specifies tolerance.
+/// * `matrix_max_num_qubits`: The maximum number of qubits to use for more expensive
+///   matrix-based checks.
+fn commute(
+    dag: &DAGCircuit,
+    inst1: &PackedInstruction,
+    inst2: &PackedInstruction,
+    approximation_degree: f64,
+    matrix_max_num_qubits: u32,
+    commutation_checker: &mut CommutationChecker,
+) -> bool {
+    let qargs1 = dag.get_qargs(inst1.qubits);
+    let qargs2 = dag.get_qargs(inst2.qubits);
+    let cargs1 = dag.get_cargs(inst1.clbits);
+    let cargs2 = dag.get_cargs(inst2.clbits);
+
+    if !qargs1.iter().any(|e| qargs2.contains(e)) && !cargs1.iter().any(|e| cargs2.contains(e)) {
+        return true;
+    }
+
+    let op1 = inst1.op.view();
+    let op2 = inst2.op.view();
+    let params1 = inst1.params_view();
+    let params2 = inst2.params_view();
+
+    commutation_checker
+        .commute(
+            &op1,
+            params1,
+            qargs1,
+            cargs1,
+            &op2,
+            params2,
+            qargs2,
+            cargs2,
+            u32::MAX,
+            matrix_max_num_qubits,
+            approximation_degree,
+        )
+        .expect("Commutation checker should work")
```

Can we guarantee it should work? If yes, maybe leave a comment why, and if not we should propagate the error 🙂

---

You called out my laziness, handling the result in [6678169](https://github.com/Qiskit/qiskit/pull/15047/commits/66781690d2b1a7dff320d1f53b383b9d87f555ad).